x/bsonx/bsoncore/bsonbuilder_test.go

```diff
@@ -0,0 +1,291 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncore
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math"
+	"reflect"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+func emptyDocumentBuilder() *DocumentBuilder {
+	return &DocumentBuilder{}
+}
+
+func emptyArrayBuilder() *ArrayBuilder {
+	a := ArrayBuilder{}
+	a.indexes = append(a.indexes, int32(0))
+	a.key = append(a.key, 0)
+	return &a
+}
+
+func TestDocumentBuilder(t *testing.T) {
+	bits := math.Float64bits(3.14159)
+	pi := make([]byte, 8)
+	binary.LittleEndian.PutUint64(pi, bits)
+
+	testCases := []struct {
+		name     string
+		fn       interface{}
+		params   []interface{}
+		expected []byte
+	}{
+		{
+			"AppendInt32",
+			emptyDocumentBuilder().AppendInt32,
+			[]interface{}{"foobar", int32(256)},
+			[]byte{byte(bsontype.Int32), 'f', 'o', 'o', 'b', 'a', 'r', 0x00, 0x00, 0x01, 0x00, 0x00},
```

Comment:
You can probably make this easier on yourself by using the existing `bsoncore` functions and building a full document like:

```
bsoncore.BuildDocumentFromElements(nil, bsoncore.AppendInt32Element(nil, "foobar", 256))
```

Those functions should already have tests, so we don't need to build raw BSON ourselves here.